darts/dataprocessing/pipeline.py

```diff
@@ -0,0 +1,166 @@
+        For each data transformer in pipeline first fit the data if transformer is fittable then transform data using
+        fitted transformer. The transformed data is then passed to next transformer.
+
+        Parameters
+        ----------
+        data
+            TimeSeries to fit and transform on.
+
+        Returns
+        -------
+        TimeSeries
+            Transformed data.
+        """
+        for transformer in self._transformers:
+            if isinstance(transformer, FittableDataTransformer):
+                transformer.fit(data)
+
+            data = transformer.transform(data)
+        return data
+
+    def transform(self, data: TimeSeries) -> TimeSeries:
+        """
+        For each data transformer in pipeline transform data. Then transformed data is passed to next transformer.
+
+        Parameters
+        ----------
+        data
+            TimeSeries to be transformed.
+
+        Returns
+        -------
+        TimeSeries
+            Transformed data.
+        """
+        for transformer in self._transformers:
+            data = transformer.transform(data)
+        return data
+
+    def inverse_transform(self, data: TimeSeries) -> TimeSeries:
+        """
+        For each data transformer in pipeline inverse_transform data. Then inverse transformed data is passed to next
+        transformer. Transformers are traversed in reverse order. Raises value error if not all of the transformers are
+        invertible.
+
+        Parameters
+        ----------
+        data
+            TimeSeries to be inverse transformed.
+
+        Returns
+        -------
+        TimeSeries
+            Inverse transformed data.
+        """
+        raise_if_not(self._invertible, "Not all transformers in the pipeline can perform inverse_transform", logger)
+
+        for transformer in reversed(self._transformers):
+            data = transformer.inverse_transform(data)
+        return data
+
+    def __getitem__(self, key: Union[int, slice]) -> 'Pipeline':
+        """
+        Gets subset of Pipeline based either on index or slice with indexes.
+        Resulting pipeline will deep copy transformers of the original pipeline.
+
+        Parameters
+        ----------
+        key
+            Either int or slice indicating the subset of data transformers to keep.
+
+        Returns
+        -------
+        Pipeline
+            Subset of pipeline determined by key.
+        """
+        raise_if_not(isinstance(key, int) or isinstance(key, slice), "key must be either an int or a slice", logger)
+
+        if isinstance(key, int):
+            transformers = [self._transformers[key]]
+        else:
+            transformers = self._transformers[key]
+        return Pipeline(transformers, deep=True)
+
+    def __iter__(self) -> Iterator[BaseDataTransformer[TimeSeries]]:
+        """
+        Returns
+        -------
+        Iterator
+            List of data transformers
+        """
+        return self._transformers
```

Comment:
I think that's an `Iterable` (and not an `Iterator`).

Reply:
I just looked, and to make `Pipeline` _iterable_ it needs to provide an `__iter__` method which returns an `Iterator`. The types are correct, but the return value isn't. I'll fix it.

Reply:
Maybe something like `return._transformers.__iter__()`

Reply:
Oops I thought I had done exactly this, sorry!